### DISCUSSION TITLE
BLDC-tool connects to vesc, teleop doesn’t

### DISCUSSION THREAD

**Discussion by drewbus-drewbus** [prl-mushr/mushr#42] [Q&A] [answered]
Hey everyone! I installed from scratch in a 4gb jetson nano and am experimenting with using ros2 and ros1 together. I had previously installed on a 2gb nano using the mushr image and everything worked great. 

Now this time I'm having drive issues. I'm getting the "Failed to connect to the VESC, SerialException Failed to open the serial port to the VESC. IO Exception (2): No such file or directory" error. However, I have the bldc-tool running on the same nano and can connect to and operate the vesc just fine. It has the same settings I downloaded and used previously and successfully in the 2gb nano. 

I'm out of ideas. I had to chmod ttyACM0 to get the bldc-tool to work and hoped permissions might fix the issue, but it doesn't appear to. I'm using a bench power supply so power should be fine. Im not sourcing ros2 - just melodic. I'm stumped. Ideas?

**Comment by palanc** [comment, marked as answer]
Hi @drewbus-drewbus - just to clarify, when you say that you installed from scratch, does that mean that you started with the stock Jetson Nano image and then used our install script process (https://mushr.io/hardware/nano_install/) to install everything? Or did you try to install stuff using your own process?

**Reply by drewbus-drewbus** [reply]
Good question. I followed the QuickStart install instructions.

**Reply by palanc** [reply]
I see, could you try the following:

1. Plug in the VESC.
2. Turn on the Jetson Nano
3. Successfully connect to the VESC using bldc-tool
4. Close bldc-tool
5. Enter the following terminal command and take a screenshot of the output:
     `ls /dev/`
6. Try running teleop again

Then could you please post the screenshot and also let us know if you still get the  'Failed to connect to the VESC...' error when launching teleop?

**Reply by palanc** [reply]
Sorry actually ignore my previous comment for now. The quickstart instructions (I assume you're referring to this: https://mushr.io/tutorials/quickstart/) are meant for installing the MuSHR simulation on a desktop/laptop rather than installing on the actual car. Could you try installing with the install script (https://mushr.io/hardware/nano_install/) instead?

**Reply by drewbus-drewbus** [reply]
Reading through the correct install script and seeing the udev rules I went "there it is." Everything works again. Thanks!

**Reply by rogeriobonatti** [reply]
@drewbus-drewbus can you please comment on what exactly you did to solve this issue? I ran the mushr installation script and all the udev rules are there (and everything seems correct), but for some weird reason the rules are not being applied to the system. I tried re-starting the rules several times but doesn't work (even after I ran some parts of the installation script manually).

**Reply by drewbus-drewbus** [reply]
I just ran the relevant parts of the hardware install script that pertained to the vesc. My issue was I'd used the simulator install script without realizing it. 

Have you added your user to the appropriate group? I can't remember which one offhand, it's in the script.

**Reply by rogeriobonatti** [reply]
Actually I figured out my issue: the lidar script was not running because the ROS IP is set to the UW IP address from their install script by default, so ROS was simply not starting on my computer. It wasn’t obvious from the launch file, but I caught it after running roscore by itself. The install script worked fine otherwise!

**Comment by hardikbvaghela** [comment]
i have the same problem. when i connect VESC tool with BLDC Tool. and then teleoperation with Plazstation remote control. but the error comes with the VESC Tool. so now wha should i do?
![IMG-3868](https://user-images.githubusercontent.com/93780099/158998322-d7ab7d47-4123-46c3-b8bb-18c068f2321f.jpg)

**Reply by palanc** [reply]
Hi @hardikbvaghela - I see that you are getting this error in **BLDC Tool**. Have you first uploaded the correct firmware in **VESC Tool**? Specifically, have you gone through this section starting at **step 18**: [https://mushr.io/hardware/build_instructions/#vesc-preparation](https://mushr.io/hardware/build_instructions/#vesc-preparation)

**Reply by hardikbvaghela** [reply]
thank you. now BLDC tool is connected but there are some issue with operating when i press forward key from keyboard but wheels turn backward and when i press backward key or left right key then wheel just jurking not working properly. so what should i do?

**Reply by palanc** [reply]
For the issue with the wheels spinning backwards when you press the up arrow on the keyboard, you can just swap any pair of motor-vesc connections in order to make the wheels turn in the correct direction. 

With respect to the wheels not turning when pressing the left or right key, it may not be a problem. I would double check that the BLDC motor was installed properly, i.e. redo **step 17** of this section: [https://mushr.io/hardware/build_instructions/#brushless-motor-installation](https://mushr.io/hardware/build_instructions/#brushless-motor-installation). As long as that step goes fine, it should be fine to continue with the build.

**Reply by hardikbvaghela** [reply]
Thank you for your response. 
Teleoperate the Robot
In this section, we will go through the minimum process required to teleoperate the MuSHR Racecar.

Steps:
Open a terminal and log into the robot:

mushr01@mushr01-desktop:~$ ssh robot@192.168.0.15 -X
robot@192.168.0.15's password: 
Permission denied, please try again.
robot@192.168.0.15's password: 
Permission denied, please try again.
robot@192.168.0.15's password: 
robot@192.168.0.15: Permission denied (publickey,password).

i put my wifi connection password it can not work i put my robot login password it can not work always give me error. 
so now what should i do?

**Reply by palanc** [reply]
The password should be **prl_robot**, have you tried that?

**Reply by hardikbvaghela** [reply]
ya its work 
and setup all the things but the last error when i run teleop.launch , my lidar is spinning but teleoperating is not working with remote control.

[FATAL] [1649101680.192804663]: Failed to connect to the VESC, SerialException Failed to open the serial port to the VESC. IO Exception (2): No such file or directory, file /tmp/binarydeb/ros-melodic-serial-1.2.1/src/impl/unix.cc, line 151. failed..
[car/vesc/vesc_driver-5] process has finished cleanly
log file: /home/mushr01/.ros/log/1b007022-b450-11ec-ae94-f8597151e929/car-vesc-vesc_driver-5*.log
[car/vesc/vesc_driver-5] restarting process
process[car/vesc/vesc_driver-5]: started with pid [5772]
[FATAL] [1649101680.612680903]: Failed to connect to the VESC, SerialException Failed to open the serial port to the VESC. IO Exception (2): No such file or directory, file /tmp/binarydeb/ros-melodic-serial-1.2.1/src/impl/unix.cc, line 151. failed..
[car/vesc/vesc_driver-5] process has finished cleanly
log file: /home/mushr01/.ros/log/1b007022-b450-11ec-ae94-f8597151e929/car-vesc-vesc_driver-5*.log
[car/vesc/vesc_driver-5] restarting process
process[car/vesc/vesc_driver-5]: started with pid [5803]
[FATAL] [1649101680.935025291]: Failed to connect to the VESC, SerialException Failed to open the serial port to the VESC. IO Exception (2): No such file or directory, file /tmp/binarydeb/ros-melodic-serial-1.2.1/src/impl/unix.cc, line 151. failed..
[car/vesc/vesc_driver-5] process has finished cleanly
log file: /home/mushr01/.ros/log/1b007022-b450-11ec-ae94-f8597151e929/car-vesc-vesc_driver-5*.log
[car/vesc/vesc_driver-5] restarting process
process[car/vesc/vesc_driver-5]: started with pid [5832]

continuously this error shows.

**Reply by palanc** [reply]
A few things could be going wrong:

1. Is the VESC battery charged?
2. Did you plug in the VESC battery before turning on the Jetson Nano? (You should)
3. After plugging in the VESC to the battery, its red LED should flash 3 times. Then the red LED should stay off and you should a constantly on blue led. Is that what you see?

**Reply by hardikbvaghela** [reply]
1,yes
2, yes
3, yes
everthing what you told 3 points like samething is working.

**Reply by palanc** [reply]
Did you compile your workspace using ```catkin_make``` before launching teleop.launch?

Were you able to successfully complete **steps 10-12** of the software setup section? [https://mushr.io/hardware/build_instructions/#software-setup](https://mushr.io/hardware/build_instructions/#software-setup)

**Reply by hardikbvaghela** [reply]
yes i make also catkin_make  and 2nd thing yes BLDC tool is completely fine it is working and connected.

**Reply by palanc** [reply]
Could you provide a screenshot of the output of executing  ```ls /dev``` in the terminal when the VESC is plugged in?

**Reply by hardikbvaghela** [reply]
it shows no such file or directory

**Reply by palanc** [reply]
So you have no directory called ```/dev```? In that case, I would suggest that you try reinstalling the OS again: [https://mushr.io/hardware/nano_install/](https://mushr.io/hardware/nano_install/)

**Reply by hardikbvaghela** [reply]
sorry sorry i got this when i write ls /dev
![IMG_4593](https://user-images.githubusercontent.com/93780099/161635355-51e16734-cbc7-4ea3-8d7d-36ae6f6f7db2.jpg)
![IMG_4594](https://user-images.githubusercontent.com/93780099/161635371-0f221007-a21d-4478-a123-3c095454490e.jpg)

**Reply by palanc** [reply]
There should be a symbolic link named 'vesc' in the /dev directory, but appears to be missing for your car. Could you execute the following commands on your robot:

1. ```sudo adduser $USER dialout```
2. ```echo "ACTION==\"add\", ATTRS{idVendor}==\"0483\", ATTRS{idProduct}==\"5740\", SYMLINK+=\"vesc\"" > /tmp/10-vesc.rules```
3. ```sudo mv /tmp/10-vesc.rules /etc/udev/rules.d/```
4. ```sudo udevadm control --reload-rules```
5.  ```sudo udevadm trigger```

Then make sure the VESC is powered and reboot the Jetson Nano. Then try launching teleop.launch again and see if the error goes away. Also please let us know if the 'vesc' symbolic link shows up in /dev (it should be colored cyan, similar to 'ydlidar').

**Reply by hardikbvaghela** [reply]
still not work VESC link is not shows in /dev. still its  same fatal error

**Reply by hardikbvaghela** [reply]
can i try everything from the 1st step? 2nd time flash the memory card and all things?

**Reply by hardikbvaghela** [reply]
because in this it gives alot of times error.

**Reply by palanc** [reply]
That's very strange that the symbolic link doesn't show up. Were there any errors when you executed the commands I listed in my previous message?

A hack that you could try to get around it is to open this file and change ```port: /dev/vesc``` to ```port: /dev/ttyACM0```: [https://github.com/prl-mushr/vesc/blob/master/vesc_main/config/racecar-uw-nano/vesc.yaml#L27](https://github.com/prl-mushr/vesc/blob/master/vesc_main/config/racecar-uw-nano/vesc.yaml#L27)

Or as you said, you could try reflashing the SD card using these instructions: [https://mushr.io/hardware/nano_install/](https://mushr.io/hardware/nano_install/)

**Reply by hardikbvaghela** [reply]
Thank you. lidar is working and now error gone everything is working cool. but i was connected mushr in HDMI cable with my display. so how can i wireless connect , or how can i run only mushr with out HDMI cable ?

**Reply by palanc** [reply]
Awesome, could you share what made it work? Was it modifying the vesc.yaml file, re-flashing the SD card, or something else?

**Step 8** here has instructions for setting up a wireless connection: [https://mushr.io/hardware/nano_install/](https://mushr.io/hardware/nano_install/)

You can then follow these steps to start teleoperating the robot through a remote connection: [https://mushr.io/hardware/build_instructions/#teleoperate-the-robot](https://mushr.io/hardware/build_instructions/#teleoperate-the-robot)

**Reply by hardikbvaghela** [reply]
no, just I follow your solution and suddenly it starts lidar is spinning and the wheel is also operated by PS4 remote controller.
and now you gave me a link for teleoperation. but my question is if I pull out the HDMI cable from the mushr ok so I just control the robot means forward-backward and all but how can I operate lidar also? how can I completely without HDMI run my robot with lidar spinning on road with the help of Remote controller?

**Reply by palanc** [reply]
I don't understand what you are asking. If you want to launch the lidar (i.e. make it spin and publish data) and be able to control the robot with the PS4 controller without using an HDMI cable, I would suggest doing the steps I laid out in my previous message. If it's something else that is the issue, could you please clarify?

**Reply by hardikbvaghela** [reply]
hello, my MuSHR is connected via HDMI cable with my other display monitor so I can run any command so I just write in my monitor screen terminal. ok. if I write relaunch teleop. launch. the lidar is spinning and the robot is actual with my ps4 controller. ok but now if I want to run my robot and also lidar spinning with connectin of HDMI cable with my monitor. so how it is possble? for the command writing where can i write relaunch teleop. launch? because there is no connection with HDMI so how can I actuate robot wireless? 

I think you understand my question. my focus is to without an HDMI connection how can I run mushr with spinning lidar and getting the measurements.

**Reply by palanc** [reply]
Okay I believe the steps I listed two messages ago will give you what you want, but I will list them out more explicitly here so that it is hopefully clearer.

1. With your current setup (HDMI cable plugged into Nano/Monitor), follow **step 8** here to get the robot to broadcast its own network: [https://mushr.io/hardware/nano_install/](https://mushr.io/hardware/nano_install/). 
2. Shutdown the robot (for example, you could use ```sudo shutdown -P now```). Remove the HDMI cable from the Nano (as well as any peripherals such as mouse/keyboard). Unpower the Nano and VESC.
3. Follow the instructions here to launch the robot remotely: [https://mushr.io/hardware/build_instructions/#teleoperate-the-robot](https://mushr.io/hardware/build_instructions/#teleoperate-the-robot). Note that in **step 4**, **Robot AP** is a network that should show up in your laptop/desktop's list of available WIFI networks to connect to. You should connect to this network so that you can ssh into the robot in the subsequent steps.

**Reply by WensongVincent** [reply]
Hi @palanc, I got the same problem of "The firmware on the connected VESC is too old, please update it using a programmer". But I've implemented the step 18 for VESC prepare. The VESC I'm using is in this link: [https://www.amazon.com/Flipsky-Aluminum-Skiing-Electric-Controller/dp/B087M4CRQX/ref=dp_prsubs_sccl_2/135-4997271-8027165?pd_rd_w=cJg0g&content-id=amzn1.sym.7df936a4-0c8f-4c2c-ae82-092487072d30&pf_rd_p=7df936a4-0c8f-4c2c-ae82-092487072d30&pf_rd_r=H10VNB9A3G1V79WX8HKH&pd_rd_wg=OxM6V&pd_rd_r=7b3d2a98-9049-43f2-9d0e-d86b78bad906&pd_rd_i=B087M4CRQX&psc=1](url), it is also a Flipsky FSESC 4.12 VESC. Could you please help me out?

**Reply by palanc** [reply]
Hi @WensongVincent - Are you sure that you **successfully** uploaded the firmware through VESC tool? Step 21 of VESC preparation states that a post-firmware dialog will pop up once upload is complete, did that happen for you?

**Reply by WensongVincent** [reply]
Hi @palanc, thank you for your reply. I've fixed the problem by flash the default firmware (version 6.2) first and then flash the one provided (version 2.18).
But now I got a new problem that, when I power up the VESC, it automatically steer to left and constantly apply force, in addition, the car is moving forward when I press left or right key on keyboard, the up and down key only making some noice, and the car can not turn left or right. Do you know what should I do? I've tried to switch the connection of motor and server, but didn't solve it.

**Reply by palanc** [reply]
For the servo, did you install the arm as shown in Fig 5.2 of the instructions ([https://mushr.io/hardware/build_instructions/#servo-motor-installation](https://mushr.io/hardware/build_instructions/#servo-motor-installation))? If it is installed offset from that position then there could be mismatch between where the robot thinks the servo arm is and where it really is.

Just to be clear, pressing the arrow keys (both up/down and left/right) only controls the main BLDC motor. None of the arrow keys control the servo. It may be that the up/down key don't cause the BLDC motor to move if the motor gears are not well meshed. Did you verify that the gears were well meshed with step 17 of the BLDC motor install ([https://mushr.io/hardware/build_instructions/#brushless-motor-installation](https://mushr.io/hardware/build_instructions/#brushless-motor-installation))
